include/swift/AST/DiagnosticsParse.def

```diff
@@ -1644,10 +1644,9 @@ ERROR(redundant_class_requirement,none,
       "redundant 'class' requirement", ())
 ERROR(late_class_requirement,none,
       "'class' must come first in the requirement list", ())
-ERROR(where_without_generic_params,none,
-      "'where' clause cannot be attached to "
-      "%select{a non-generic|a protocol|an associated type}0 "
-      "declaration", (unsigned))
+ERROR(where_toplevel_nongeneric,none,
+      "'where' clause cannot be attached to non-generic "
+      "top-level declaration", ())
```

Comment:
Why not diagnose this in Sema?

Reply:
Because we are able to safely diagnose this earlier without additional semantic information, no?

Reply:
What about the `where` clause though? If we diagnose in Parse, the where clause is simply not added to the declaration, unlike Sema, where it can trigger additional unnecessary routines and possibly mess up other diagnostics code, unless we already have some `isInvalid` flag for where clauses. Do you think it is worth a shot anyway?

Reply:
It's up to you, but I prefer to diagnose related issues in one place. Nothing should look at the where clause except for the code that builds the declaration's generic signature, which is where you can diagnose invalid where clauses.

Reply:
Makes sense, I will give it a try!